### PR TITLE
*: update rust toolchain to 2022-01-07

### PR DESCRIPTION
**Pull request by tabokie**
Signed-off-by: tabokie <xy.tao@outlook.com>

### What is changed and how it works?

Ref #11119 

What's Changed:

Some rust packages have opt in edition2021, which is stable after rust 1.56. E.g. https://github.com/tikv/raft-engine/pull/178

### Check List 

Tests
- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by Connor1996 — COMMENTED** [review]
no need to change the edition to `2021`?

**Comment by sticnarf** [comment]
> 

I think we can postpone changing TiKV itself to 2021 later. Changing to 2021 may involve some closure capture changes.

File: components/test_storage/src/sync_storage.rs [hunk 1]
```diff
@@ -154,8 +156,8 @@ impl<E: Engine> SyncTestStorage<E> {
     ) -> Result<Vec<Option<Vec<u8>>>> {
         let mut ids = vec![];
         let requests: Vec<GetRequest> = keys
-            .to_owned()
-            .into_iter()
+            .iter()
+            .copied()
```
Comment (Connor1996) [inline components/test_storage/src/sync_storage.rs:160]: what's the difference between `copied` and `cloned`?

**Review by Connor1996 — COMMENTED** [review]
rest LGTM

File: components/raftstore/src/store/worker/region.rs [hunk 2]
```diff
@@ -806,7 +806,7 @@ mod tests {
         // which are [f, i), [m, n), [p, t)
         let timeout2 = 12;
         let overlap_ranges =
-            pending_delete_ranges.drain_overlap_ranges(&b"g".to_vec(), &b"q".to_vec());
+            pending_delete_ranges.drain_overlap_ranges(b"g".as_ref(), b"q".as_ref());
```
Comment (sticnarf) [inline components/raftstore/src/store/worker/region.rs:809]:
```suggestion
            pending_delete_ranges.drain_overlap_ranges(b"g", b"q");
```

Comment (sticnarf) [reply components/raftstore/src/store/worker/region.rs:809]: All similar `as_ref`s in the PR seem not necessary.

**Review by sticnarf — COMMENTED** [review]
The rest lgtm

File: components/test_storage/src/sync_storage.rs (hunk 1 above)
Comment (tabokie) [reply components/test_storage/src/sync_storage.rs:160]: I think clippy prefers `copied` over `cloned`, but sometimes `Copy` is not implemented (for `Vec`).

**Review by youjiali1995 — APPROVED** [review]

**Review by sticnarf — APPROVED** [review]

File: src/coprocessor/statistics/cmsketch.rs [hunk 3]
```diff
@@ -6,6 +6,7 @@ use murmur3::murmur3_x64_128;
 /// Refer:[Count-Min Sketch](https://en.wikipedia.org/wiki/Count-min_sketch)
 #[derive(Clone)]
 pub struct CmSketch {
+    #[allow(dead_code)]
```
Comment (BusyJay) [inline src/coprocessor/statistics/cmsketch.rs:9]: Can these be deleted?

**Comment by tabokie** [comment]
Somehow the newer rust starts to report an integer underflow here:
https://github.com/tikv/tikv/blob/6170255604aff4a52b365c91d09859e4315045c5/components/cdc/src/channel.rs#L390
@overvenus Could you help triage?

File: src/coprocessor/statistics/cmsketch.rs (hunk 3 above)
Comment (tabokie) [reply src/coprocessor/statistics/cmsketch.rs:9]: In this case, the `depth` is used in tests, so I'll keep it. Will clean up other cases.

**Comment by tabokie** [comment]
@BusyJay Spotted another case of underflow:
https://github.com/tikv/tikv/blob/6170255604aff4a52b365c91d09859e4315045c5/components/raftstore/src/store/fsm/store.rs#L863
I'm not sure if it could be related to #11111. It doesn't look very harmful so I'll just mute it for now.

**Comment by sticnarf** [comment]
Seems there is a regression https://github.com/rust-lang/rust/issues/93054 about generating code coverage. It may affect https://github.com/tikv/tikv/issues/10949

**Comment by tabokie** [comment]
@sticnarf Changed the version to 2022-01-07.

**Comment by sticnarf** [comment]
> Changed the version to 2022-01-07.

I have no dissent but am just curious is it a randomly chosen toolchain or is there any special reason?

**Comment by tabokie** [comment]
> > Changed the version to 2022-01-07.
> 
> I have no dissent but am just curious is it a randomly chosen toolchain or is there any special reason?

Am pretty sure 7 is everyone's favorite number 🙃

**Pull request merged**